### ISSUE TITLE
service name not available for the specified socket type

### ISSUE THREAD

**Issue by mtelka** [eventlet/eventlet#791]
I tested 0.33.3 on OpenIndiana with Python 3.9.16 and I found that five tests failed with errors like these:
```
ERROR: test_canonname (tests.greendns_test.TestGetaddrinfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "$(BUILD_DIR)/tests/greendns_test.py", line 687, in test_canonname
    res = greendns.getaddrinfo('host.example.com', 0,
  File "$(BUILD_DIR)/eventlet/support/greendns.py", line 558, in getaddrinfo
    ai = socket.getaddrinfo(addr, port, family,
  File "/usr/lib/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 9] service name not available for the specified socket type
```
The problem is that `port=0` is passed to `socket.getaddrinfo`.  Here is simple reproducer:
```
$ python -c 'import socket; socket.getaddrinfo("google.com", 0)' ; echo $?
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 9] service name not available for the specified socket type
1
$ python -c 'import socket; socket.getaddrinfo("google.com", None)' ; echo $?
0
$ python -c 'import socket; socket.getaddrinfo("google.com", 80)' ; echo $?
0
$
```
When I test this on Linux all three commands above succeeds, so apparently the difference is in platform.

The only unclear thing to me is whether the actual problem is in the platform (i.e. `socket.getaddrinfo(port=0)` should succeed, if not it is a platform bug), or in eventlet tests (i.e. calling `socket.getaddrinfo(port=0)` is not guaranteed to work, so it should be avoided).

I looked at the `socket.getaddrinfo` documentation at https://docs.python.org/3/library/socket.html#socket.getaddrinfo.  There is documented that `port=STRING` should work, also `port=NUMBER` should work, and finally `port=None` is documented too.  For other `socket.getaddrinfo` parameters (like family, type and proto) it is also specified that passing zero there is okay, but not for `port`.  So it looks like it is not guaranteed that `socket.getaddrinfo(port=0)` must work, so I assume eventlet's tests should be updated.  Probably to pass `port=None` instead of `port=0`.

Thank you.

**Comment by temoto** [comment]
please post all test failures

**Comment by temoto** [comment]
please try this fix dd2f0ea

**Comment by mtelka** [comment]
> please try this fix [dd2f0ea](https://github.com/eventlet/eventlet/commit/dd2f0ea32d6b6d064be52fe8e27fca275985cdc4)

I tried the fix and now the following tests pass:
```
test_canonname (tests.greendns_test.TestGetaddrinfo)
test_getaddrinfo_hosts_only_ans_with_error (tests.greendns_test.TestGetaddrinfo)
test_getaddrinfo_hosts_only_ans_with_timeout (tests.greendns_test.TestGetaddrinfo)
test_getaddrinfo_only_a_ans (tests.greendns_test.TestGetaddrinfo)
test_getaddrinfo_only_aaaa_ans (tests.greendns_test.TestGetaddrinfo)
```

Thank you!

**Comment by temoto** [comment]
Fix is merged in master, release is coming after some testing.

FYI using any means to announce, next 2 minor releases are last to support python <3.5 (drop-py27 is coming soon).
